File: offer/003.倒序输出链表/list.cpp
/**
*  struct ListNode {
*        int val;
*        struct ListNode *next;
*        ListNode(int x) :
*              val(x), next(NULL) {
*        }
*  };
*/
class Solution {
public:
    vector<int> printListFromTailToHead(ListNode* head) {
        ListNode *dummy = new ListNode(-1), *tail = dummy;
        ListNode *p = head;
        int len = 0;
        while(p != nullptr){
            ListNode *next = p->next;
            p->next = dummy->next;
            dummy->next = p;
            p = next;
            len++;
        }
        //cout << "len:" << len << endl;
        vector<int> ret(len);
        p = dummy->next;
        int index = 0;
        while(p != nullptr){
            ret[index] = p->val;
            p = p->next;
            index++;
        }
        return ret;
    }
};

File: offer/JZ6 从尾到头打印链表/list.java
/**
*    public class ListNode {
*        int val;
*        ListNode next = null;
*
*        ListNode(int val) {
*            this.val = val;
*        }
*    }
*
*/
import java.util.ArrayList;
public class Solution {
    public ArrayList<Integer> printListFromTailToHead(ListNode head) {
        ListNode dummy = new ListNode(-1);
        while(head != null){
            ListNode old = dummy.next, p = head;
            head = head.next;
            
            dummy.next = p;
            p.next = old;
            
        }
        ArrayList<Integer> ret = new ArrayList<>();
        dummy = dummy.next;
        while(dummy != null){
            ret.add(dummy.val);
            dummy = dummy.next;
        }
        return ret;
    }
}